core/src/com/tgf/twf/core/world/storage/HashMapInventory.java

```diff
@@ -41,25 +84,61 @@ public boolean store(final Inventory inventory) {
         return true;
     }
 
+    @Override
+    public boolean reserve(final ResourceType resourceType, final int quantity) {
+        if (quantity == 0) {
+            return true;
+        }
+        final Stock stock = stocks.get(resourceType);
+        if (stock == null) {
+            stocks.put(resourceType, Stock.ofReservedQuantity(quantity));
+        } else {
+            stock.reservedQuantity += quantity;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean reserve(final Inventory inventory) {
+        inventory.getStoredResourceTypes().forEach(
+                resourceType -> reserve(resourceType, inventory.getStoredQuantity(resourceType))
+        );
+        return true;
+    }
 
     @Override
     public boolean retrieve(final ResourceType resourceType, final int quantity) {
-        final int currentStock = stock.getOrDefault(resourceType, 0);
-        final int newStock = currentStock - quantity;
+        if (quantity == 0) {
+            return true;
+        }
+        final Stock stock = stocks.get(resourceType);
+        if (null == stock) {
+            return false;
+        }
+        final int newStock = stock.storedQuantity - quantity;
         if (newStock < 0) {
-            throw new IllegalStateException("Cannot retrieve " + quantity + " of " + resourceType + "; only " + currentStock + " available");
+            return false;
         }
-        if (newStock == 0) {
-            stock.remove(resourceType);
+        if (newStock == 0 && stock.reservedQuantity == 0) {
+            stocks.remove(resourceType);
         } else {
-            stock.put(resourceType, newStock);
+            stock.storedQuantity = newStock;
         }
         totalStoredQuantity -= quantity;
-        return newStock == 0;
+        return true;
+    }
+
+    @Override
+    public boolean retrieve(final Inventory inventory) {
+        inventory.getStoredResourceTypes().forEach(
```

Comment:
I think for inventories containing multiple resource types this operation is not atomic. Do we need a `canRetrieve` followed by a `retrieve` here? Or is this the responsibility of the caller?

Reply:
Same comment as above. 
See the Storage methods:

    /**
     * @param resources The quantity of resources to consume.
     * @return true if the storage have enough resources in store to satisfy the demand.
     */
    public boolean canRetrieve(final Inventory resources) {
        return resources.getStoredResourceTypes().stream()
                .allMatch(resourceType -> inventory.getStoredQuantity(resourceType) >= resources.getStoredQuantity(resourceType));
    }

    /**
     * @param resources The quantity of resources to consume. {@link #canRetrieve(Inventory)} should be called first to check if the storage can
     *                  satisfy the demand or not. Insufficient capacity will throw an {@link IllegalStateException}.
     */
    public void retrieve(final Inventory resources) {
        resources.getStoredResourceTypes()
                .forEach(resourceType -> {
                    if (!inventory.retrieve(resourceType, resources.getStoredQuantity(resourceType))) {
                        throw new IllegalStateException("Storage rejected retrieval of " + resourceType);
                    }
                });
    }